core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java

```diff
@@ -758,6 +758,8 @@ public PlanOptimizers(
                             // Must run before AddExchanges and after ReplicateSemiJoinInDelete
                             // to avoid temporarily having an invalid plan
                             new DetermineSemiJoinDistributionType(costComparator, taskCountEstimator)))));
+            // Try to push into table scan again (specifically joins) since ReorderJoins and DetermineJoinDistributionType have run
+            builder.add(pushIntoTableScanOptimizer);
```

Comment:
why in `!forceSingleNode` section?

Reply:
Doesn't matter I guess as flag is only used for `LocalQueryRunner`

Reply:
Because of the `AddExchanges` applied before exiting this `if` block.

Reply:
https://github.com/trinodb/trino/pull/7257 is adding a `DetermineTableScanNodePartitioning` rule that determines whether table scan node partitioning will be used. That rule must be executed before `AddExchanges` and at least once during planning since plan shape and fragmenting depends on that rule.

`pushIntoTableScanOptimizer` cannot be executed after `DetermineTableScanNodePartitioning` because returned table handle might change node partitioning (e.g table handle might point to a different table with different properties).

Could we move `pushIntoTableScanOptimizer` just after reorder joins (I'm not sure `DetermineJoinDistributionType` really matters here).

Otherwise https://github.com/trinodb/trino/pull/7257 should land first

Reply:
> Otherwise #7257 should land first

What does this one buys us, apart from "other person have to worry" ? :)

Reply:
This PR in current form logically conflicts with https://github.com/trinodb/trino/pull/7257. Given that https://github.com/trinodb/trino/pull/7257 was created earlier and approved (and is much larger), it should land first IMO

Reply:
Yeah sure. I do not object :) Just was wondering what you meant.

Reply:
https://github.com/trinodb/trino/pull/7257 landed